<a href="https://colab.research.google.com/github/Bharat745/H2O/blob/master/H2O_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Final Project

## Installation and Imports

In [0]:
# Installing java version for running H2O on colab
# If code is run on Jupyter notebook, you do NOT need to run this code
! apt-get install default-jre
!java -version

Reading package lists... Done
Building dependency tree       
Reading state information... Done
default-jre is already the newest version (2:1.11-68ubuntu1~18.04.1).
0 upgraded, 0 newly installed, 0 to remove and 28 not upgraded.
openjdk version "11.0.4" 2019-07-16
OpenJDK Runtime Environment (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3)
OpenJDK 64-Bit Server VM (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3, mixed mode, sharing)


In [0]:
# Start and connect to local H2O cluster
# If the code is run on Jupyter notebook, remove the '!' in pip install
! pip install h2o
import h2o
h2o.init(nthreads = -1)

Checking whether there is an H2O instance running at http://localhost:54321 ..... not found.
Attempting to start a local H2O server...
  Java Version: openjdk version "11.0.4" 2019-07-16; OpenJDK Runtime Environment (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3); OpenJDK 64-Bit Server VM (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3, mixed mode, sharing)
  Starting server from /usr/local/lib/python3.6/dist-packages/h2o/backend/bin/h2o.jar
  Ice root: /tmp/tmpgfhkhovu
  JVM stdout: /tmp/tmpgfhkhovu/h2o_unknownUser_started_from_python.out
  JVM stderr: /tmp/tmpgfhkhovu/h2o_unknownUser_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321 ... successful.


--------------------------  ------------------------------------------------------------------
H2O cluster uptime:         02 secs
H2O cluster timezone:       Etc/UTC
H2O data parsing timezone:  UTC
H2O cluster version:        3.26.0.8
H2O cluster version age:    8 days
H2O cluster name:           H2O_from_python_unknownUser_lkyyjz
H2O cluster total nodes:    1
H2O cluster free memory:    2.938 Gb
H2O cluster total cores:    2
H2O cluster allowed cores:  2
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:       {'http': None, 'https': None}
H2O internal security:      False
H2O API Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, TargetEncoder, Core V4
Python version:             3.6.8 final
--------------------------  ------------------------------------------------------------------

In [0]:
# Importing matplotlib to make plots
import pandas as pd
import matplotlib.pyplot as plt
import datetime as dt
%matplotlib inline

In [0]:
# Importing the data file
df_h2o = h2o.import_file("http://coursera.h2o.ai/house_data.3487.csv")

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [0]:
# Converting the h2o frame to pandas data frame
df_pd = df_h2o.as_data_frame(use_pandas = True)

In [0]:
df_pd.head(5)

id             date   price  ...     long  sqft_living15  sqft_lot15
0  7129300520  20141013T000000  221900  ... -122.257           1340        5650
1  6414100192  20141209T000000  538000  ... -122.319           1690        7639
2  5631500400  20150225T000000  180000  ... -122.233           2720        8062
3  2487200875  20141209T000000  604000  ... -122.393           1360        5000
4  1954400510  20150218T000000  510000  ... -122.045           1800        7503

[5 rows x 21 columns]

In [0]:
# Splitting the date column into Date and Time at "T"
new = df_pd['date'].str.split("T", n=1, expand =True)
df_pd["Date"]= new[0]
df_pd["Time"] = new[1]

# Dropping the original Date column
df_pd.drop(columns =["date"], inplace = True)

df_pd.head(5)

id   price  bedrooms  ...  sqft_lot15      Date    Time
0  7129300520  221900         3  ...        5650  20141013  000000
1  6414100192  538000         3  ...        7639  20141209  000000
2  5631500400  180000         2  ...        8062  20150225  000000
3  2487200875  604000         4  ...        5000  20141209  000000
4  1954400510  510000         3  ...        7503  20150218  000000

[5 rows x 22 columns]

In [0]:
# Converting the new Date column to month and year

from datetime import datetime
date_format = df_pd['Date']
df_pd['date1'] = df_pd['Date'].apply(lambda x: datetime.strptime(x, '%Y%m%d'))

df_pd["month"] = df_pd['date1'].map(lambda x: x.month)
df_pd["year"] = df_pd['date1'].map(lambda x: x.year)



df_pd.drop(columns =["Date", "Time", "date1"], inplace = True)



In [0]:
# Converting the pandas data frame back to h2o frame

hf = h2o.H2OFrame(df_pd)

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [0]:
hf.head(5)

id price bedrooms bathrooms sqft_living sqft_lot floors waterfront view condition grade sqft_above sqft_basement yr_built yr_renovated zipcode lat long sqft_living15 sqft_lot15 month year 
 
 
 7.1293e+09 221900 3 1 1180 5650 1 0 0 3 7 1180 0 1955 0 98178 47.5112 -122.257 1340 5650 10 2014 
 6.4141e+09 538000 3 2.25 2570 7242 2 0 0 3 7 2170 400 1951 1991 98125 47.721 -122.319 1690 7639 12 2014 
 5.6315e+09 180000 2 1 770 10000 1 0 0 3 6 770 0 1933 0 98028 47.7379 -122.233 2720 8062 2 2015 
 2.4872e+09 604000 4 3 1960 5000 1 0 0 5 7 1050 910 1965 0 98136 47.5208 -122.393 1360 5000 12 2014 
 1.9544e+09 510000 3 2 1680 8080 1 0 0 3 8 1680 0 1987 0 98074 47.6168 -122.045 1800 7503 2 2015

## Data Preparation

In [0]:
# Splitting the data to train, validation & test
train,test = hf.split_frame([0.9] , seed = 123)

In [0]:
print(train.nrows)
print(test.nrows)

19462
2151


In [0]:
# Setting the model parameters
ignoreFields = [
    "id" , "zipcode", "price"
    ]

x =[i for i in train.names if i not in ignoreFields]

y = "price"


In [0]:
print("Response = " + y)
print("Predictors = " + str(x))

Response = price
Predictors = ['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'lat', 'long', 'sqft_living15', 'sqft_lot15', 'month', 'year']


## Modelling / Analysis

In [0]:
# Importing all the necessary model parameters
from h2o.estimators.random_forest import H2ORandomForestEstimator
from h2o.estimators.gbm import H2OGradientBoostingEstimator
from h2o.estimators.glm import H2OGeneralizedLinearEstimator
from h2o.estimators.deeplearning import H2ODeepLearningEstimator
from h2o.estimators.stackedensemble import H2OStackedEnsembleEstimator
from h2o.grid.grid_search import H2OGridSearch

In [0]:
# defining the model and setting the parameters
nfolds = 5

The main aim of the experiment is to predict the price of the house using different models and the aim is to bring the RMSE value below 123,000. 

### GLM Model

In [0]:
# GLM Model
m_GLM = H2OGeneralizedLinearEstimator(
    model_id = "glm_def", 
    nfolds = nfolds,
    fold_assignment = "Modulo",
    keep_cross_validation_predictions = True
  )
m_GLM.train(x, y, training_frame = train)

glm Model Build progress: |███████████████████████████████████████████████| 100%


In [0]:
m_GLM

Model Details
H2OGeneralizedLinearEstimator :  Generalized Linear Modeling
Model Key:  glm_def


GLM Model: summary


family  ... number_of_iterations training_frame
0    gaussian  ...                    1  py_4_sid_97c4

[1 rows x 8 columns]



ModelMetricsRegressionGLM: glm
** Reported on train data. **

MSE: 132394522442.04964
RMSE: 363860.58105000824
MAE: 230692.2794816973
RMSLE: 0.540677127127496
R^2: 0.01915498238848845
Mean Residual Deviance: 132394522442.04964
Null degrees of freedom: 19461
Residual degrees of freedom: 19442
Null deviance: 2626981989511147.5
Residual deviance: 2576662195767170.0
AIC: 553676.1355983858

ModelMetricsRegressionGLM: glm
** Reported on cross-validation data. **

MSE: 132917725402.59248
RMSE: 364578.8329053025
MAE: 231261.72157556852
RMSLE: 0.5419978801831691
R^2: 0.015278832472447434
Mean Residual Deviance: 132917725402.59248
Null degrees of freedom: 19461
Residual degrees of freedom: 19442
Null deviance: 2627276624027082.0
Residual deviance: 2586844771785255.0
AIC: 553752.8948833451

Cross-Validation Metrics Summary: 


mean  ...     cv_4_valid     cv_5_valid
0                     mae      231261.72  ...      231744.22      232231.47
1  mean_residual_deviance  1.32917166E11  ...  1.33148361E11  1.22194231E11
2                     mse  1.32917166E11  ...  1.33148361E11  1.22194231E11
3           null_deviance  5.25455325E14  ...  5.26203388E14  4.83281095E14
4                      r2    0.015090036  ...    0.015025707   0.0155544095
5       residual_deviance  5.17368942E14  ...  5.18213406E14   4.7557995E14
6                    rmse      364468.44  ...       364895.0       349562.9
7                   rmsle      0.5419561  ...      0.5412729      0.5507107

[8 rows x 8 columns]


Scoring History: 


timestamp  ... negative_log_likelihood     objective
0    2019-10-26 20:20:52  ...            2.626982e+15  1.349801e+11

[1 rows x 6 columns]

As we can see from the above model, that RMSE is 363860.58 on Training data and 364578.83 on Test data. Hence we can say that the GLM model does not perform as good.

### GBM Model

In [0]:
# GBM Model
m_GBM = H2OGradientBoostingEstimator( 
    model_id = "gbm_def", 
    nfolds = nfolds,
    fold_assignment = "Modulo",
    keep_cross_validation_predictions = True
  )

m_GBM.train(x, y, training_frame = train)

gbm Model Build progress: |███████████████████████████████████████████████| 100%


In [0]:
m_GBM

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  gbm_def


Model Summary: 


number_of_trees  ...  max_leaves  mean_leaves
0               50.0  ...        31.0        26.46

[1 rows x 10 columns]



ModelMetricsRegression: gbm
** Reported on train data. **

MSE: 11678437596.824286
RMSE: 108066.82005511352
MAE: 67662.44394284375
RMSLE: 0.17619086287094732
Mean Residual Deviance: 11678437596.824286

ModelMetricsRegression: gbm
** Reported on cross-validation data. **

MSE: 17063623560.686049
RMSE: 130627.80546532215
MAE: 74189.4899406734
RMSLE: 0.18637404579733474
Mean Residual Deviance: 17063623560.686049

Cross-Validation Metrics Summary: 


mean  ...     cv_4_valid     cv_5_valid
0                     mae       74189.28  ...        72551.8       72710.39
1  mean_residual_deviance  1.70634598E10  ...  1.73712609E10  1.47183278E10
2                     mse  1.70634598E10  ...  1.73712609E10  1.47183278E10
3                      r2      0.8739623  ...      0.8714949      0.8814233
4       residual_deviance  1.70634598E10  ...  1.73712609E10  1.47183278E10
5                    rmse      130369.33  ...      131800.08      121319.12
6                   rmsle     0.18632275  ...     0.18489002     0.18559027

[7 rows x 8 columns]


Scoring History: 


timestamp  ...   training_mae  training_deviance
0     2019-10-26 20:21:10  ...  233426.255360       1.349801e+11
1     2019-10-26 20:21:10  ...  215267.599598       1.160429e+11
2     2019-10-26 20:21:10  ...  199406.066612       1.003873e+11
3     2019-10-26 20:21:10  ...  184701.101125       8.728549e+10
4     2019-10-26 20:21:10  ...  172075.276646       7.630885e+10
5     2019-10-26 20:21:10  ...  160860.229147       6.729103e+10
6     2019-10-26 20:21:10  ...  150577.599455       5.939710e+10
7     2019-10-26 20:21:11  ...  141987.154641       5.299554e+10
8     2019-10-26 20:21:11  ...  133911.266865       4.741331e+10
9     2019-10-26 20:21:11  ...  127018.605815       4.287064e+10
10    2019-10-26 20:21:11  ...  120725.930562       3.895019e+10
11    2019-10-26 20:21:11  ...  115230.898398       3.558126e+10
12    2019-10-26 20:21:11  ...  110522.917130       3.288403e+10
13    2019-10-26 20:21:11  ...  106063.734505       3.043271e+10
14    2019-10-26 20:21:11  ...  102504.320965       2.839701e+10
15    2019-10-26 20:21:11  ...   99010.240891       2.656941e+10
16    2019-10-26 20:21:11  ...   96077.138414       2.508417e+10
17    2019-10-26 20:21:11  ...   93267.179817       2.371172e+10
18    2019-10-26 20:21:11  ...   91058.268649       2.248148e+10
19    2019-10-26 20:21:11  ...   88864.956629       2.149335e+10

[20 rows x 7 columns]


See the whole table with table.as_data_frame()

Variable Importances: 


variable  relative_importance  scaled_importance  percentage
0           grade         4.320886e+15           1.000000    0.342115
1     sqft_living         3.445133e+15           0.797321    0.272775
2             lat         2.057736e+15           0.476230    0.162925
3            long         9.996391e+14           0.231351    0.079148
4      waterfront         4.133098e+14           0.095654    0.032725
5   sqft_living15         3.923390e+14           0.090801    0.031064
6        yr_built         3.214826e+14           0.074402    0.025454
7            view         2.317232e+14           0.053629    0.018347
8      sqft_above         2.014532e+14           0.046623    0.015950
9        sqft_lot         6.082480e+13           0.014077    0.004816
10      bathrooms         5.027807e+13           0.011636    0.003981
11     sqft_lot15         3.881067e+13           0.008982    0.003073
12      condition         3.019198e+13           0.006987    0.002391
13  sqft_basement         2.123633e+13           0.004915    0.001681
14           year         1.261404e+13           0.002919    0.000999
15   yr_renovated         1.079940e+13           0.002499    0.000855
16         floors         1.007999e+13           0.002333    0.000798
17          month         8.656172e+12           0.002003    0.000685
18       bedrooms         2.730922e+12           0.000632    0.000216

As we can see from the above model that RMSE value for both the training and the Cross validation set is above the required value. From all the Cross-validation results, only CV5 is below the required value. overall the model is better than the GLM Model.

### Random Forest Model

In [0]:
# RF Model
m_RF = H2ORandomForestEstimator( 
    model_id = "rf_def", 
    nfolds = nfolds,
    fold_assignment = "Modulo",
    keep_cross_validation_predictions = True
  )

m_RF.train(x, y, training_frame = train)

drf Model Build progress: |███████████████████████████████████████████████| 100%


In [0]:
m_RF

Model Details
H2ORandomForestEstimator :  Distributed Random Forest
Model Key:  rf_def


Model Summary: 


number_of_trees  ...  max_leaves  mean_leaves
0               50.0  ...      9816.0      8962.26

[1 rows x 10 columns]



ModelMetricsRegression: drf
** Reported on train data. **

MSE: 17924626859.477493
RMSE: 133882.8848638895
MAE: 72708.06608446881
RMSLE: 0.18228542377203544
Mean Residual Deviance: 17924626859.477493

ModelMetricsRegression: drf
** Reported on cross-validation data. **

MSE: 17655253706.369812
RMSE: 132873.07366946025
MAE: 71782.93647892366
RMSLE: 0.1808673346909977
Mean Residual Deviance: 17655253706.369812

Cross-Validation Metrics Summary: 


mean  ...     cv_4_valid     cv_5_valid
0                     mae       71782.71  ...       70468.19       69211.52
1  mean_residual_deviance  1.76550216E10  ...  1.67388621E10  1.50702746E10
2                     mse  1.76550216E10  ...  1.67388621E10  1.50702746E10
3                      r2     0.86968964  ...      0.8761731     0.87858784
4       residual_deviance  1.76550216E10  ...  1.67388621E10  1.50702746E10
5                    rmse      132543.98  ...      129378.75      122761.05
6                   rmsle     0.18083099  ...     0.18143064      0.1782299

[7 rows x 8 columns]


Scoring History: 


timestamp  ...   training_mae  training_deviance
0     2019-10-26 20:22:04  ...            NaN                NaN
1     2019-10-26 20:22:04  ...  112653.976359       4.456665e+10
2     2019-10-26 20:22:04  ...  107181.227473       3.845702e+10
3     2019-10-26 20:22:05  ...  102173.218939       3.533720e+10
4     2019-10-26 20:22:05  ...   98474.811347       3.204043e+10
5     2019-10-26 20:22:05  ...   95850.260158       3.085767e+10
6     2019-10-26 20:22:05  ...   93188.065481       2.905689e+10
7     2019-10-26 20:22:05  ...   91312.044748       2.789338e+10
8     2019-10-26 20:22:06  ...   89628.402530       2.677217e+10
9     2019-10-26 20:22:06  ...   88203.798068       2.599982e+10
10    2019-10-26 20:22:06  ...   87088.348050       2.571881e+10
11    2019-10-26 20:22:06  ...   85343.472943       2.434710e+10
12    2019-10-26 20:22:06  ...   83996.166934       2.310312e+10
13    2019-10-26 20:22:07  ...   82762.794492       2.229706e+10
14    2019-10-26 20:22:07  ...   81811.833443       2.185337e+10
15    2019-10-26 20:22:07  ...   80919.969458       2.152561e+10
16    2019-10-26 20:22:07  ...   80231.993958       2.115459e+10
17    2019-10-26 20:22:07  ...   79642.951720       2.114559e+10
18    2019-10-26 20:22:07  ...   78699.947313       2.049337e+10
19    2019-10-26 20:22:08  ...   78111.837231       2.026615e+10

[20 rows x 7 columns]


See the whole table with table.as_data_frame()

Variable Importances: 


variable  relative_importance  scaled_importance  percentage
0           grade         2.365917e+16           1.000000    0.245299
1     sqft_living         2.214670e+16           0.936072    0.229618
2             lat         1.235343e+16           0.522141    0.128081
3      sqft_above         8.658763e+15           0.365979    0.089774
4   sqft_living15         7.087009e+15           0.299546    0.073478
5            long         5.378614e+15           0.227337    0.055766
6       bathrooms         3.253917e+15           0.137533    0.033737
7        yr_built         3.005582e+15           0.127037    0.031162
8      waterfront         2.269444e+15           0.095922    0.023530
9            view         2.120620e+15           0.089632    0.021987
10  sqft_basement         1.533520e+15           0.064817    0.015900
11       sqft_lot         9.614498e+14           0.040638    0.009968
12     sqft_lot15         9.449252e+14           0.039939    0.009797
13          month         7.874878e+14           0.033285    0.008165
14       bedrooms         6.845079e+14           0.028932    0.007097
15         floors         5.715890e+14           0.024159    0.005926
16      condition         5.191022e+14           0.021941    0.005382
17   yr_renovated         3.322269e+14           0.014042    0.003445
18           year         1.821733e+14           0.007700    0.001889

As we see from the Random Forest Model that the RMSE on both the train and cross-validation set is above the required range.

### Deep Learning Model

In [0]:
# Deep Learning Model
m_DL = H2ODeepLearningEstimator(
    model_id ="dl_def", 
    nfolds = nfolds,
    fold_assignment = "Modulo",
    keep_cross_validation_predictions = True
  )

m_DL.train(x, y, training_frame = train)

deeplearning Model Build progress: |██████████████████████████████████████| 100%


In [0]:
m_DL

Model Details
H2ODeepLearningEstimator :  Deep Learning
Model Key:  dl_def


Status of Neuron Layers: predicting price, regression, gaussian distribution, Quadratic loss, 44,401 weights/biases, 530.3 KB, 201,589 training samples, mini-batch size 1


layer  units       type  ... mean_weight weight_rms  mean_bias      bias_rms
0        1     19      Input  ...                                                
1        2    200  Rectifier  ...   0.0010875   0.130017   0.443833     0.0872253
2        3    200  Rectifier  ...  -0.0210937  0.0865518   0.963565     0.0263995
3        4      1     Linear  ...   0.0116019  0.0727976  0.0612793  1.09713e-154

[4 rows x 14 columns]



ModelMetricsRegression: deeplearning
** Reported on train data. **

MSE: 12832674068.281149
RMSE: 113281.39330128822
MAE: 69033.53196619314
RMSLE: 0.1787971876990994
Mean Residual Deviance: 12832674068.281149

ModelMetricsRegression: deeplearning
** Reported on cross-validation data. **

MSE: 15369154238.018976
RMSE: 123972.39304788376
MAE: 71954.90968878994
RMSLE: 0.1820919909045422
Mean Residual Deviance: 15369154238.018976

Cross-Validation Metrics Summary: 


mean  ...     cv_4_valid    cv_5_valid
0                     mae      71954.875  ...       71618.15      71121.44
1  mean_residual_deviance  1.53690962E10  ...  1.57833257E10  1.3706538E10
2                     mse  1.53690962E10  ...  1.57833257E10  1.3706538E10
3                      r2     0.88626677  ...      0.8832417    0.88957465
4       residual_deviance  1.53690962E10  ...  1.57833257E10  1.3706538E10
5                    rmse      123851.11  ...      125631.71     117074.92
6                   rmsle     0.18208386  ...     0.18331435    0.18059652

[7 rows x 8 columns]


Scoring History: 


timestamp  ...   training_mae training_r2
0    2019-10-26 20:25:00  ...            NaN         NaN
1    2019-10-26 20:25:06  ...  110111.292725    0.784365
2    2019-10-26 20:25:14  ...   80005.818598    0.864257
3    2019-10-26 20:25:20  ...   91206.693009    0.859039
4    2019-10-26 20:25:26  ...   72348.891972    0.892888
5    2019-10-26 20:25:33  ...   75307.181723    0.894243
6    2019-10-26 20:25:39  ...   70562.432247    0.905102
7    2019-10-26 20:25:41  ...   69033.531966    0.905376

[8 rows x 11 columns]


Variable Importances: 


variable  relative_importance  scaled_importance  percentage
0             lat             1.000000           1.000000    0.139329
1            long             0.885940           0.885940    0.123437
2        yr_built             0.381632           0.381632    0.053172
3        sqft_lot             0.366994           0.366994    0.051133
4      sqft_lot15             0.363674           0.363674    0.050670
5   sqft_living15             0.349313           0.349313    0.048669
6           grade             0.349002           0.349002    0.048626
7     sqft_living             0.343475           0.343475    0.047856
8      sqft_above             0.324302           0.324302    0.045185
9          floors             0.297282           0.297282    0.041420
10      condition             0.296009           0.296009    0.041243
11          month             0.295436           0.295436    0.041163
12       bedrooms             0.294786           0.294786    0.041072
13           year             0.293721           0.293721    0.040924
14     waterfront             0.292103           0.292103    0.040698
15      bathrooms             0.271149           0.271149    0.037779
16  sqft_basement             0.270581           0.270581    0.037700
17   yr_renovated             0.260643           0.260643    0.036315
18           view             0.241232           0.241232    0.033611

We see from all the above models that Random Forest model performs the best among all. The RMSE for Training set is 115284.27 while on cross-validation set is 125447.43. We get the RMSE value below the required range on the test set but not on the CV set. Hence, we decide to move forward with Stacked Ensemble.

### Stacked Ensemble

In [0]:
# we create a list of 3 model ids
models = [m_GBM.model_id, m_RF.model_id, m_DL.model_id]

In [0]:
m_SE = H2OStackedEnsembleEstimator(model_id = "SE_models",
                           base_models = models)
m_SE.train(x, y, train)

stackedensemble Model Build progress: |███████████████████████████████████| 100%


In [0]:
m_SE

Model Details
H2OStackedEnsembleEstimator :  Stacked Ensemble
Model Key:  SE_models

No model summary for this model

ModelMetricsRegressionGLM: stackedensemble
** Reported on train data. **

MSE: 8370660436.407039
RMSE: 91491.31344781886
MAE: 56502.960833979
RMSLE: 0.1471052124374389
R^2: 0.937985949632007
Mean Residual Deviance: 8370660436.407039
Null degrees of freedom: 19461
Residual degrees of freedom: 19458
Null deviance: 2626981989511147.5
Residual deviance: 162909793413353.78
AIC: 499908.51267742383


In [0]:
all_models = [m_GBM, m_RF,m_DL, m_SE]
names = ["GBM", "RF", "DL", "SE"]

In [0]:
pd.Series(map(lambda x: x.rmse(), all_models), names)

GBM    108066.820055
RF     133882.884864
DL     113281.393301
SE      91491.313448
dtype: float64

As we can see from the above results, the Stacked Ensemble performs the best with RMSE of 93072.63 which is way below the required range. Our Stacked Ensemble Model outperforms all the other models. 

### Checking the performance on test data 

In [0]:
m_GLM.model_performance(test)


ModelMetricsRegressionGLM: glm
** Reported on test data. **

MSE: 130259499571.50009
RMSE: 360914.80929923075
MAE: 235461.0434662482
RMSLE: 0.5493152576219195
R^2: 0.020090090321806087
Mean Residual Deviance: 130259499571.50009
Null degrees of freedom: 2150
Residual degrees of freedom: 2131
Null deviance: 285935013037402.56
Residual deviance: 280188183578296.7
AIC: 61196.374428120056


In [0]:
m_GBM.model_performance(test)


ModelMetricsRegression: gbm
** Reported on test data. **

MSE: 17840789803.042454
RMSE: 133569.41941568232
MAE: 74849.92191914232
RMSLE: 0.18370543141484513
Mean Residual Deviance: 17840789803.042454


In [0]:
m_RF.model_performance(test)


ModelMetricsRegression: drf
** Reported on test data. **

MSE: 17222346237.15325
RMSE: 131233.93706337264
MAE: 71596.07334415679
RMSLE: 0.1768392780064867
Mean Residual Deviance: 17222346237.15325


In [0]:
m_DL.model_performance(test)


ModelMetricsRegression: deeplearning
** Reported on test data. **

MSE: 15866073096.986208
RMSE: 125960.60136799207
MAE: 73501.66422579202
RMSLE: 0.18632793700343367
Mean Residual Deviance: 15866073096.986208


In [0]:
m_SE.model_performance(test)


ModelMetricsRegressionGLM: stackedensemble
** Reported on test data. **

MSE: 14948081525.550259
RMSE: 122262.34712923787
MAE: 68938.95048658423
RMSLE: 0.17220282596881048
R^2: 0.8875492899500662
Mean Residual Deviance: 14948081525.550259
Null degrees of freedom: 2150
Residual degrees of freedom: 2147
Null deviance: 285935013037402.56
Residual deviance: 32153323361458.605
AIC: 56507.5763440207


When we check individual model performance on the test data, we see that Stacked Ensemble Model performs as expected giving us RMSE below the target value.

In [0]:
# Saving the SE Model

# save the model
model_path = h2o.save_model(model=m_SE, path="C:/Users/bkhandhadia2/OneDrive - DXC Production/H2O_course/Models", force=True)

In [0]:
print (model_path)

/content/C:/Users/bkhandhadia2/OneDrive - DXC Production/H2O_course/Models/SE_models


In [0]:
h2o.shutdown()

[WARNING] in <ipython-input-37-1edf85295eae> line 1:
    >>> h2o.shutdown()
        ^^^^ Deprecated, use ``h2o.cluster().shutdown()``.
H2O session _sid_97c4 closed.
